[#307] MyBundlePage v1.2.0

kim-hyunjoo opened the pull request:
<!--제목 템플릿-->
<!--[#1] 프로젝트 세팅 v1.0.0-->
# 📝작업 내용
### 1. 전체 레이아웃 수정
![image](https://github.com/Team-kiwing/Team-3seco-kiwing-fe/assets/78135416/2a7e55db-27ce-4b5a-aefc-e6fabf713127)

- 팀에서 논의한 레이아웃 규격에 맞도록 기존 레이아웃을 수정했습니다.
- 내 꾸러미 페이지는 UI가 많기 때문에, 불가피하게 width: 95%로 적용하게 되었습니다.

### 2. UI개선
![image](https://github.com/Team-kiwing/Team-3seco-kiwing-fe/assets/78135416/037c0390-4af3-414e-99b0-0e0ffce96628)
![image](https://github.com/Team-kiwing/Team-3seco-kiwing-fe/assets/78135416/ff4c9fe3-86b4-48f4-b496-238ad894f5b9)
- 꾸러미, 질문이 없을 때 그냥 비어있는 기존 UI가 사용자에게 불친절하다고 느껴졌습니다.
- 따라서 사용자로 하여금 액션을 유도하는 UI 컴포넌트를 구현했습니다.
- 몇몇 컴포넌트는 BorderBox -> ShadowBox로 다시 변경했습니다.

### 3. react-query 리패치 최적화
- 기존에는 질문에 변경사항이 있을 시, `queryClient.refetchQueries({ queryKey: [QUERYKEY.BUNDLE_DETAIL] })` 로 전체 꾸러미에 대해 불필요한 api호출을 실행하게 되며 성능 저하 이슈를 감지했습니다.
- 따라서 queryKey 배열에 bundleId를 같이 넣어 해당 꾸러미에 대한 리패치만 이루어지도록 개선했습니다.

### 4. 잘못된 주소 리다이렉트
https://github.com/Team-kiwing/Team-3seco-kiwing-fe/assets/78135416/2f76c84e-7e3b-4358-a91b-0298736c96ec
- 내 꾸러미 페이지는 유저의 아이디가 주소에 들어갑니다.
- 따라서 사용자가 주소에 임의로 다른 값을 넣어도, 올바른 주소로 다시 돌아갈 수 있도록 리다이렉트해주는 로직을 MyBundlePage 상단에 구현했습니다.

### 5. 드래그 앤 드롭 개선
https://github.com/Team-kiwing/Team-3seco-kiwing-fe/assets/78135416/072eab53-3993-4593-9b1a-dd39357a1e8d
- 기존에는 QuestionBox의 모든 곳이 다 draggable 했기 때문에 답변을 복사할 수 없는 이슈, 모바일 환경에서는 중첩 드래그 이벤트 발생하는 이슈가 발생했습니다.
- 따라서 각 컴포넌트의 헤더를 드래그 할 때만 움직이도록 수정하여 해당 이슈를 해결했습니다.

## 다음 개선사항 TODO

- [x] textarea 줄바꿈 안되는 현상 해결
- [x] 토글 위치와 공개 여부 UI 개선 -> 툴팁 컴포넌트 구현되는대로 개선
- [ ] api status를 통해 정확하게 분기 처리하여 에러 핸들링 -> 팀원 모두가 같은 api 함수를 사용하기 때문에 제가 함부로 먼저 바꿀 수 없을 것 같습니다. 우선 같은 함수를 사용하고 있는 팀원이 있는지 확인 한 후에 있다면 같이 수정해야 할 것 같습니다.
- [ ] 전체적인 코드 구조 리팩토링 : 상수화, 공통 로직 분리
- [x] refetch 최적화 안된 곳 전부 적용하기
- [x] contenteditable 사용했을 때 answer 낙관적 업데이트

# ✨PR Point
- 자잘하게 UI가 추가되고 바뀌었는데 현재 UI 괜찮은지 여쭤보고 싶습니다!
- api 함수 status로 정확히 분기하는건 service의 api함수를 `return res.data` -> `return res` 로 변경해야하기 때문에 아무래도 다같이 수정을 해야할 것 같습니다 ㅜㅜ

kim-hyunjoo committed:
🐛fix : key 중첩 삭제

kim-hyunjoo committed:
✨feat : 낙관적 업데이트 추가, 모바일에서 isEditMode일 때 height값 보정

kim-hyunjoo committed:
♻️refactor : 모바일에서 detail query 선택된 것만 refetch하도록 개선

kim-hyunjoo committed:
♻️refactor : console.log 제거

kim-hyunjoo committed:
💄design : BorderBox -> ShadowBox로 변경, 레이아웃 비율 변경

- ShadowBox로 변경하며 그림자가 어색하게 잘리는 현상이 생겼습니다.
- 이에 따라 전체 container width: 90%->95%로 변경, padding 값을 부여했습니다.

kim-hyunjoo committed:
💄design : borderbox->shadowbox로 변경하며 어색해진 css 수정

kim-hyunjoo committed:
✨feat : 내 꾸러미 페이지에 다른 주소가 들어와도 다시 사용자 아이디 주소로 리다이렉트

- 사용자가 임의로 주소에 다른값을 넣어도 사용자id를 주소로 갖도록 리다이렉트 시킵니다.
- 로딩 스피너를 적용했습니다.

kim-hyunjoo committed:
✨feat : 꾸러미에 질문이 없는 경우 안내 UI 추가

- 꾸러미에 질문이 없는 경우, 질문을 추가할 수 있도록 유도하는 UI를 제작했습니다.
- 다른 사람의 질문을 가져올 수 있도록 유도하기 위해 허브 이동 링크를 넣었습니다.

kim-hyunjoo committed:
💄design : 질문이 없을 때 안내 UI gap부여, 화살표 애니메이션 적용

kim-hyunjoo committed:
💄design : 화살표 색 theme gray_300으로 변경

kim-hyunjoo committed:
💄design : 화살표 아이콘 cursor: auto로 변경

kim-hyunjoo committed:
✨feat : index 컴포넌트에 question이 없는 경우 안내 UI 추가

- 질문이 없을 땐 질문을 추가되면 목차가 생긴다는 안내 문구 추가
- 맨 위에 목차를 의미하는 이모지 추가

kim-hyunjoo committed:
✨feat : 꾸러미가 없을 때 안내 UI 컴포넌트 구현

kim-hyunjoo committed:
♻️refactor : 꾸러미가 없을 때 안내 UI 나오도록 구조 변경

- MyBundleDetail props에서 isMyBundleEmpty 제거
- 꾸러미 비어있는 경우 MyBudleEmpty 컴포넌트 보이도록 페이지에서 분기 처리
- 어색한 css 개선

kim-hyunjoo committed:
💄design : MyBundleEmpty ShadowBox 삭제, 모바일 환경 css 개선

- ShadowBox를 삭제하고 Container에서 css를 개선했습니다.
- 모바일 환경에서는 버튼이 하단에 고정되도록 변경했습니다.

kim-hyunjoo committed:
♻️refactor : MyQuestionEmpty 컴포넌트로 분리

- 기존 MyBundleDetail 내부의 안내 UI를 새로운 컴포넌트로 분리했습니다.

kim-hyunjoo committed:
💄design : MyBundleEmpty, MyQuestionEmpty 텍스트 추가 및 변경

- 꾸러미가 비어있는 경우에도 질문이 비어있을 때 처럼 더욱 상세하게 텍스트 추가
- 꾸러미, 질문 둘 다 텍스트 내용 자연스럽게 변경

kim-hyunjoo committed:
💄design : MyBundleEmpty, MyQuestionEmpty 회색 변경

- border_color -> gray_300으로 변경했습니다.

kim-hyunjoo committed:
🐛fix : 질문이 비어있는 조건을 filteredQuestions->bundle.questions로 변경

kim-hyunjoo committed:
🐛fix : 질문 헤더만 드래그 되도록 수정

- 기존에는 답변을 복사하기 위해 드래그를 하면 드래그 앤 드롭 이벤트가 발동되었기 때문에
이에 사용자가 불편함을 느끼게 되었습니다.
- 따라서 질문 Header에 dragHandleProps를 부여하여 이 문제를 해결했습니다.

kim-hyunjoo committed:
♻️refactor : BundleDetail 쿼리 키 bundleId string->number타입으로 수정

kim-hyunjoo committed:
♻️refactor : 누락된 bundleId 타입 변경

kim-hyunjoo committed:
✨feat : 모바일에서 중첩 드래그 안 되도록 구조 개선

kim-hyunjoo committed:
💄design : Empty 컴포넌트 키윙 로고 이미지 크기 변경

kim-hyunjoo committed:
🐛fix : 모바일 사이즈에서 꾸러미 삭제 후 웹으로 돌아왔을 때 삭제한 꾸러미 api호출하는 현상 해결

kim-hyunjoo committed:
✨feat : MyBundleMenu 스켈레톤 UI 추가

kim-hyunjoo committed:
♻️refactor : 낙관적 업데이트에 오류가 있어 다시 롤백

- 모달로 답변을 수정하는 경우 업데이트가 안되는 현상 발견

kim-hyunjoo committed:
♻️refactor : QuestionBox bundleId 선택 속성으로 변경

pkb9239 commented on src/components/MyBundle/MyBundleEmpty/MyBundleEmpty.style.ts:
```diff
@@ -0,0 +1,111 @@
+`;
+
+export const Text = styled.span`
+  font-size: ${WEB_FONT_SIZE}rem;
```
P5; 글자색 `primary_color` 쓰셔야할거같습니다.

kim-hyunjoo replied:
어이쿠 다크모드를 생각 못했네요!!! 이 부분 추가하겠습니다 감사합니다 ㅎㅎ

pkb9239 commented on src/components/MyBundle/MyBundleEmpty/MyBundleEmpty.style.ts:
```diff
@@ -0,0 +1,111 @@
+    }
+  }
+
+  @keyframes skeleton--loading {
```
오오 !

kim-hyunjoo replied:
경빈님 덕분에 아주 유용하게 썼습니닿ㅎ....👍👍👍

pkb9239 reviewed (approved):
엄청나게 사용자가 사용하기 편해진거같습니다 ! ! 가독성이 좋아지고 친절해진거같습니다. 수고많으셨습니다. 너무 이쁩니다 !

kim-hyunjoo committed:
✨feat : watch() 이용해 답변 낙관적 업데이트

- 기존 getValues()는 모달에서 값이 바뀌었을 때 리렌더링을 일으키지 않았습니다.
- 따라서 watch()로 바꾸어 바뀐 값에 따라 리렌더링이 일어나도록 했습니다.

kim-hyunjoo committed:
💄design : BundleEmpty Text 색 primary_color 적용

- 기존에 다크모드에서도 검은색 글씨로 나와 가독성이 떨어지는 현상이 발견되었습니다.
- primary_color를 적용하여 해결했습니다.

kim-hyunjoo committed:
💄design : QuestionEmpty text-align: center 적용

kim-hyunjoo committed:
Merge branch 'dev' of https://github.com/Team-kiwing/Team-3seco-kiwing-fe into #307/myBundlePage-1.2.0

kim-hyunjoo committed:
♻️refactor : toast 위치 수정, stacked로 변경

- 기존에 toast위치가 오른쪽 위에 존재하여 몇 기능이 가려지는 현상이 발견되었습니다.
- 따라서 toast위치를 가운데 위로 조정합니다.
- toast가 여러개 뜨는 경우 UI가 가려지지 않도록 stacked 옵션을 부여했습니다.
- 모바일에서는 stacked에 문제가 있어 적용하지 않았습니다.

kim-hyunjoo committed:
🔧chore : react tooltip 설치

kim-hyunjoo committed:
✨feat : MyBundleMenu 툴팁 추가

- 기존 공개 여부, 링크 복사 기능의 경우 의미가 명확하지 않아
사용자에게 혼란을 야기할 여지가 있었습니다.
- 따라서 툴팁을 도입하여 조금 더 상세한 설명을 덧붙였습니다.

kim-hyunjoo committed:
✨feat : MyQuestionBox 토글 답변 공개 여부 관련 툴팁 추가

kim-hyunjoo committed:
💄design : QuestionBox font-weight, font-size 변경

- font-weight를 250으로 얇게 변경했습니다.
- font-size를 0.2rem씩 증가시켰습니다.

kim-hyunjoo committed:
✨feat : 꾸러미 추가 모달의 토글에 툴팁 추가, validate 강화

- 꾸러미 추가 모달의 공개 여부를 결정하는 토글에 툴팁을 적용했습니다.
- TagFilter 위에 선택을 유도하는 text를 추가했습니다.
- 기존에는 꾸러미에 빈 값을 넣어도 그대로 submit되어 api가 전송되어 400에러를 받게 되었습니다.
- validate을 강화하여 required 적용, value.trim().length===0일 때를 예외처리 했습니다.

kim-hyunjoo committed:
🐛fix : 모달로 Question 수정했을 때 QuestionBox 크기가 그대로인 현상 해결

- contenteditable로 수정했을 때는 height값을 다시 재조정해주는 로직이
있어 문제가 없었으나, 모달로 수정하는 경우에는 재조정 로직이 없어 글이 잘리는
현상이 발견되었습니다.
- useEffect를 통해 watch()를 통해 감지 중인 answer 값이 변경될 때 마다
height값을 재조정해주어 해결했습니다.

kim-hyunjoo committed:
🐛fix : isActive일 때만 height 계산하도록 변경

- isActive가 조건에 없으니 처음 꾸러미를 선택했을 때 질문들이 다 펼쳐져 보이는
오류가 발생했습니다.
- isActive 됐을 때만 계산하도록 로직을 변경했습니다.

kim-hyunjoo committed:
✨feat : MyQuestionModal 툴팁 적용 및 디자인 개선

- 기존 불친절한 UI를 툴팁 도입, placeholder 변경 등 UI를 개선했습니다.

kim-hyunjoo committed:
🐛fix : MyQuestionModal validate 강화

- 기존에는 질문을 작성하지 않아도 submit되어 api 400에러가 발생했습니다.
- validate를 강화하여 질문에 required 적용, validate를 trim()이용해
기존 조건보다 강화하여 적용했습니다.

kim-hyunjoo committed:
♻️refactor : tooltip a -> div 태그 변경

- 링크 없이 a태그를 사용하니 SEO 점수가 낮아져서 div태그로 변경했습니다.

kim-hyunjoo committed:
♻️refactor : api 분기 처리 변경, 리패치 최적화 적용

- 기존에는 api 400에러가 발생해도 onSuccess로 넘어가 성공 toast가
보이는 에러 현상이 발생했습니다.
- onSuccess에서 res로 분기처리를 한 번 더 해서 toast 내용을 분기했습니다.
- bundleDetails 관련 리패치 최적화를 추가 적용 했습니다.

kim-hyunjoo committed:
💄design : MyBundleList word-break: break-all; 적용

ksjdev reviewed (approved):
고생하셨습니다~!!

JaeHyunGround reviewed (approved):
확인했습니다 !

JaeHyunGround commented on src/components/MyBundle/MyBundleEmpty/MyBundleEmpty.tsx:
```diff
@@ -0,0 +1,74 @@
+        <Text>아직 꾸러미가 하나도 없어요😢</Text>
+        <Text>면접을 대비하기 위한 나만의 꾸러미를 생성해보세요!</Text>
+        <div>
+          <SmallText>만약 다른 사람의 꾸러미를 가져오고 싶다면?{` `}</SmallText>
```
P5; {' '} 대신 `&nbsp;` 로 대체해도 될 것 같아요 !!

kim-hyunjoo committed:
🔧chore : package-lock.json 추가

ksjdev committed:
✨feat : 툴팁 딜레이 500 -> 100으로 수정

ksjdev reviewed (approved):
킹갓엠페러충무공마제스티
메이플 300렙보다 더 대단한 갓현주
